File: backend/GraficaModerna.API/Controllers/StripeWebhookController.cs
using System.Net;
using System.Security;
using GraficaModerna.Application.Interfaces;
using GraficaModerna.Domain.Entities;
using GraficaModerna.Infrastructure.Context;
using GraficaModerna.Infrastructure.Security;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;
using Microsoft.EntityFrameworkCore;
using Stripe;
using Stripe.Checkout;

namespace GraficaModerna.API.Controllers;

[Route("api/webhook")]
[ApiController]
[AllowAnonymous]
[EnableRateLimiting("WebhookPolicy")]
public class StripeWebhookController(
    IConfiguration configuration,
    IOrderService orderService,
    ILogger<StripeWebhookController> logger,
    MetadataSecurityService securityService,
    AppDbContext context) : ControllerBase
{
    private readonly IConfiguration _configuration = configuration;
    private readonly ILogger<StripeWebhookController> _logger = logger;
    private readonly IOrderService _orderService = orderService;
    private readonly MetadataSecurityService _securityService = securityService;
    private readonly AppDbContext _context = context;

    [HttpPost("stripe")]
    public async Task<IActionResult> HandleStripeEvent()
    {
        var json = await new StreamReader(HttpContext.Request.Body).ReadToEndAsync();
        var endpointSecret = _configuration["STRIPE_WEBHOOK_SECRET"];

        if (string.IsNullOrEmpty(endpointSecret))
        {
            _logger.LogCritical("STRIPE_WEBHOOK_SECRET não configurado.");
            return StatusCode(500);
        }

        try
        {
            var signature = Request.Headers["Stripe-Signature"];
            var stripeEvent = EventUtility.ConstructEvent(json, signature, endpointSecret);

            var eventExists = await _context.ProcessedWebhookEvents
                .AsNoTracking()
                .AnyAsync(e => e.EventId == stripeEvent.Id);

            if (eventExists)
            {
                _logger.LogInformation("Evento {EventId} já processado anteriormente. Ignorando.", stripeEvent.Id);
                return Ok();
            }

            if (stripeEvent.Type == Events.CheckoutSessionCompleted)
            {
                await ProcessCheckoutSessionAsync(stripeEvent);
            }
            else if (stripeEvent.Type == Events.CheckoutSessionAsyncPaymentSucceeded)
            {
                await ProcessCheckoutSessionAsync(stripeEvent);
            }
            else if (stripeEvent.Type == Events.CheckoutSessionAsyncPaymentFailed ||
                     stripeEvent.Type == Events.PaymentIntentPaymentFailed)
            {
                _logger.LogWarning("[Webhook] Pagamento falhou: {EventId}", stripeEvent.Id);
            }
            else
            {
                _logger.LogInformation("[Webhook] Evento não tratado recebido: {EventType}. ID: {EventId}", stripeEvent.Type, stripeEvent.Id);
            }

            _context.ProcessedWebhookEvents.Add(new ProcessedWebhookEvent { EventId = stripeEvent.Id });
            await _context.SaveChangesAsync();

            return Ok();
        }
        catch (StripeException e)
        {
            _logger.LogError(e, "[Webhook] Erro de validação Stripe. Possível assinatura inválida.");
            return BadRequest("Invalid signature");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "[Webhook] Erro interno ao processar webhook {EventId}.", Request.Headers["Stripe-Event-Id"]);
            return StatusCode(500, "Internal server error");
        }
    }

    private async Task ProcessCheckoutSessionAsync(Event stripeEvent)
    {
        if (stripeEvent.Data.Object is not Session session || session.Metadata == null)
            return;

        if (session.Metadata.TryGetValue("order_data", out var encryptedOrder))
        {
            try
            {
                var plainOrderId = _securityService.Unprotect(encryptedOrder);

                if (!Guid.TryParse(plainOrderId, out var orderId))
                {
                    _logger.LogError("ID descriptografado inválido no evento {EventId}.", stripeEvent.Id);
                    throw new Exception("Invalid Order ID format");
                }

                var transactionId = session.PaymentIntentId;
                var amountPaid = session.AmountTotal ?? 0;

                if (string.IsNullOrEmpty(transactionId))
                {
                    _logger.LogError("[Webhook] PaymentIntentId ausente para Order {OrderId}. Event: {EventId}", orderId, stripeEvent.Id);
                    return;
                }

                _logger.LogInformation(
                    "[Webhook] Processando pagamento para Order {OrderId}. Transaction: {TransactionId}",
                    orderId, transactionId);

                try
                {
                    await _orderService.ConfirmPaymentViaWebhookAsync(orderId, transactionId, amountPaid);
                    _logger.LogInformation("[Webhook] Pagamento confirmado com sucesso. Order {OrderId}", orderId);
                }
                catch (Exception ex) when (ex.Message.Contains("FATAL") || ex.Message.Contains("SECURITY"))
                {
                    _logger.LogCritical(ex, "[SECURITY ALERT] Tentativa de fraude detectada. Order: {OrderId}", orderId);
                    throw;
                }
            }
            catch (SecurityException)
            {
                _logger.LogCritical("FALHA DE INTEGRIDADE: Assinatura dos metadados inválida no Webhook {EventId}.", stripeEvent.Id);
                throw;
            }
        }
        else
        {
            _logger.LogWarning("[Webhook] Metadados de segurança ausentes ou inválidos no evento {EventId}", stripeEvent.Id);
        }
    }
}

File: grafica-moderna/app/Http/Controllers/StripeWebhookController.php
<?php

namespace App\Http\Controllers\Webhook;

use App\Http\Controllers\Controller;
use App\Services\OrderService;
use App\Services\MetadataSecurityService;
use App\Models\ProcessedWebhookEvent;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;
use Stripe\Webhook;
use Stripe\Event;
use Stripe\Exception\SignatureVerificationException;

class StripeWebhookController extends Controller
{
    protected OrderService $orderService;
    protected MetadataSecurityService $securityService;

    public function __construct(
        OrderService $orderService,
        MetadataSecurityService $securityService
    ) {
        $this->orderService = $orderService;
        $this->securityService = $securityService;
    }

    public function handleStripe(Request $request)
    {
        $payload = $request->getContent();
        $signature = $request->header('Stripe-Signature');
        $endpointSecret = config('services.stripe.webhook_secret');

        if (!$endpointSecret) {
            Log::critical('STRIPE_WEBHOOK_SECRET não configurado.');
            return response()->json(null, 500);
        }

        try {
            $event = Webhook::constructEvent(
                $payload,
                $signature,
                $endpointSecret
            );

            if (ProcessedWebhookEvent::where('event_id', $event->id)->exists()) {
                Log::info("Evento {$event->id} já processado. Ignorando.");
                return response()->json(null, 200);
            }

            switch ($event->type) {
                case 'checkout.session.completed':
                case 'checkout.session.async_payment_succeeded':
                    $this->processCheckoutSession($event);
                    break;

                case 'checkout.session.async_payment_failed':
                case 'payment_intent.payment_failed':
                    Log::warning("[Webhook] Pagamento falhou: {$event->id}");
                    break;

                default:
                    Log::info("[Webhook] Evento não tratado: {$event->type} ({$event->id})");
            }

            try {
                ProcessedWebhookEvent::create([
                    'event_id' => $event->id,
                ]);
            } catch (\Throwable $e) {
                Log::warning('Race condition ao salvar webhook processado', [
                    'event_id' => $event->id,
                    'error' => $e->getMessage(),
                ]);
            }

            return response()->json(null, 200);
        }
        catch (SignatureVerificationException $e) {
            Log::error('[Webhook] Assinatura Stripe inválida.');
            return response()->json('Invalid signature', 400);
        }
        catch (\Throwable $e) {
            Log::error('[Webhook] Erro interno ao processar webhook.', [
                'exception' => $e
            ]);
            return response()->json('Internal server error', 500);
        }
    }

    protected function processCheckoutSession(Event $event): void
    {
        $session = $event->data->object;

        if (!$session || empty($session->metadata['order_data'])) {
            Log::warning("[Webhook] Metadados ausentes no evento {$event->id}");
            return;
        }

        try {
            $plainOrderId = $this->securityService->unprotect($session->metadata['order_data']);

            if (!\Ramsey\Uuid\Uuid::isValid($plainOrderId)) {
                throw new \Exception('Invalid Order ID format');
            }

            $transactionId = $session->payment_intent;
            $amountPaid = $session->amount_total ?? 0;

            if (!$transactionId) {
                Log::error("[Webhook] PaymentIntent ausente. Order {$plainOrderId}");
                return;
            }

            Log::info("[Webhook] Confirmando pagamento Order {$plainOrderId}, Transaction {$transactionId}");

            $this->orderService->confirmPaymentViaWebhook(
                $plainOrderId,
                $transactionId,
                $amountPaid
            );

            Log::info("[Webhook] Pagamento confirmado. Order {$plainOrderId}");
        }
        catch (\Exception $e) {
            Log::critical("[SECURITY ALERT] Violação de integridade no webhook {$event->id}");
            throw $e;
        }
        catch (\Throwable $e) {
            if (str_contains($e->getMessage(), 'FATAL') || str_contains($e->getMessage(), 'SECURITY')) {
                Log::critical("[SECURITY ALERT] Tentativa de fraude Order {$plainOrderId}", [
                    'exception' => $e
                ]);
                throw $e;
            }

            throw $e;
        }
    }
}